Remove Task/Coeval instances

alexandru opened the pull request:
`Local` should have side-effectful behavior, should not suspend execution, otherwise its usage is misleading.

I've added an import for synchronous behavior:

```scala
import monix.execution.misc.CanBindLocals.Implicits.synchronousAsDefault
```

What this does is to import in scope an implicit for all unknown `CanBindLocals[A]` types, but that does NOT override the implicits that are already in scope. This is important because we don't want to override the `Future` instance via an import like this.

Towards this purpose I've published a new micro library that introduces the `Not` data type, for providing evidence that an implicit does not exist for a certain type:
https://github.com/monix/implicitbox

This is a very light dependency and introducing it in Monix felt wrong.

alexandru committed:
Remove Task/Coeval instances

oleg-py commented:
Do we need to remove them? Can't we eagerly capture locals, and suspend the rest?

alexandru commented:
Don’t know, maybe.

alexandru committed:
Default implicit for CanBindLocals

alexandru commented:
@Avasil @oleg-py I've updated this branch to introduce an `CanBindLocals.Implicits.synchronousAsDefault` symbol that can be imported to provide synchronous behavior for all data types, but that does not override the implicits that are already in scope.

I've also updated the description of the PR.

alexandru committed:
Fix doc

Avasil commented on monix-execution/shared/src/main/scala/monix/execution/misc/CanBindLocals.scala:
```diff
@@ -19,84 +19,133 @@ package monix.execution.misc
+      * Implicit instance for all things synchronous.
+      *
+      * Needs to be imported explicitly in scope. Will NOT override
+      * other `CanBuildFrom` implicits that are already visible.
```
```suggestion
      * other `CanBindLocals` implicits that are already visible.
```

alexandru committed:
Fix comment

oleg-py commented:
Looks neat, let's go for it.
